src/algorithms/grassmann.jl

```diff
@@ -11,260 +11,211 @@ The module exports nothing, and all references to it should be qualified, e.g.
+        end
+        VectorInterface.scale!(x.S, abs(α))
     end
+    return x
 end
+VectorInterface.scale!!(x::GrassmannTangent, α::Number) = VectorInterface.scale!(x, α)
 
-Base.:*(g::PrecGrad, alpha::Number) = PrecGrad(g.Pg * alpha, g.g * alpha, g.rho)
-function Base.:+(a::PrecGrad, b::PrecGrad)
-    if a.rho == b.rho
-        PrecGrad(a.Pg + b.Pg, a.g + b.g, a.rho)
-    else
-        PrecGrad(a.Pg + b.Pg)
-    end
-end
-
-struct ManifoldPoint{T,E,G,C}
-    state::T # the state at that point
-    envs::E # the environments
-    g::G # the MPS gradient, which is not equivalent with the grassmann gradient!
-    Rhoreg::C # the regularized density matrices
-end
+# manifold methods
+# ----------------
+"""
+    inner(state, g1, g2)
 
-function ManifoldPoint(state::FiniteMPS, envs)
-    al_d = similar(state.AL)
-    O = envs.operator
-    for i in 1:length(state)
-        al_d[i] = MPSKit.∂∂AC(i, state, O, envs) * state.AC[i]
-    end
-    g = Grassmann.project.(al_d, state.AL)
+The inner product between tangent vectors on the left-canonical MPS manifold.
+This is given by the (Euclidean) inner product between the tangent vectors at each site.
+Because the cost function is assumed holomorphic, we take the twice the real part of the result.
 
-    Rhoreg = Vector{eltype(state.C)}(undef, length(state))
-    δmin = sqrt(eps(real(scalartype(state))))
-    tmap!(Rhoreg, 1:length(state); scheduler=MPSKit.Defaults.scheduler[]) do i
-        return regularize(state.C[i], max(norm(g[i]) / 10, δmin))
-    end
+For the effective inner product lifted from the Hilbert space metric,
+see also [`precondition`](@ref).
+"""
+inner(state, g1, g2) = 2 * real(sum(x -> Grassmann.inner(x...), zip(state.AL, g1, g2)))
 
-    return ManifoldPoint(state, envs, g, Rhoreg)
-end
-function ManifoldPoint(state::InfiniteMPS, envs)
-    δmin = sqrt(eps(real(scalartype(state))))
-    Tg = Core.Compiler.return_type(Grassmann.project,
-                                   Tuple{eltype(state.AL),eltype(state.AL)})
-    g = similar(state.AL, Tg)
-    ρ = similar(state.C)
+"""
+    precondition(state, g)
 
-    MPSKit.check_recalculate!(envs, state)
-    tforeach(1:length(state); scheduler=MPSKit.Defaults.scheduler[]) do i
-        AC′ = MPSKit.∂∂AC(i, state, envs.operator, envs) * state.AC[i]
-        g[i] = Grassmann.project(AC′, state.AL[i])
-        ρ[i] = regularize(state.C[i], max(norm(g[i]) / 10, δmin))
+In order to obtain an effective inner product between tangent vectors that is lifted from
+the inner product of the Hilbert space of the state, we additionally include the inverse
+metric: ``g ↦ g / ρ`` where `ρ` is the (regularised) right fixed point of the MPS transfer
+matrix.
+"""
+function precondition(state, g)
+    g′ = similar(g)
+    δmin = eps(real(scalartype(state)))^(3 / 4)
+    tforeach(eachindex(state); scheduler=MPSKit.Defaults.scheduler[]) do i
+        δ = max(δmin, norm(g[i]))
+        ρ = rho_regularized(state.C[i]; rtol=δ)
+        g′[i] = g[i] / ρ
```

Comment:
Since `rho_regularized` is only used here (as far as I can see), is it not easier to have it immediately compute the regularised inverse (U * (1/(S^2 + delta^2) * U') instead, thereby eliminating the inverse here.

Reply:
Great catch, this indeed should be better, and might also improve the stability slightly to ensure the inner product is positive definite at all times